presto-common/src/main/java/com/facebook/presto/common/block/AbstractMapBlock.java

```diff
@@ -0,0 +1,517 @@
+                int entryCount = entriesEndOffset - entriesStartOffset;
+
+                newOffsets[newPosition + 1] = newOffsets[newPosition] + entryCount;
+
+                for (int elementIndex = entriesStartOffset; elementIndex < entriesEndOffset; elementIndex++) {
+                    entriesPositions.add(elementIndex);
+                }
+            }
+            newPosition++;
+        }
+
+        int[] rawHashTables = getHashTables().get();
+        int[] newRawHashTables = null;
+        int newHashTableEntries = newOffsets[newOffsets.length - 1] * HASH_MULTIPLIER;
+        if (rawHashTables != null) {
+            newRawHashTables = new int[newHashTableEntries];
+            int newHashIndex = 0;
+            for (int i = offset; i < offset + length; ++i) {
+                int position = positions[i];
+                int entriesStartOffset = getOffset(position);
+                int entriesEndOffset = getOffset(position + 1);
+                for (int hashIndex = entriesStartOffset * HASH_MULTIPLIER; hashIndex < entriesEndOffset * HASH_MULTIPLIER; hashIndex++) {
+                    newRawHashTables[newHashIndex] = rawHashTables[hashIndex];
+                    newHashIndex++;
+                }
+            }
+        }
+
+        Block newKeys = getRawKeyBlock().copyPositions(entriesPositions.elements(), 0, entriesPositions.size());
+        Block newValues = getRawValueBlock().copyPositions(entriesPositions.elements(), 0, entriesPositions.size());
+        return createMapBlockInternal(
+                0,
+                length,
+                Optional.of(newMapIsNull),
+                newOffsets,
+                newKeys,
+                newValues,
+                new HashTables(Optional.ofNullable(newRawHashTables), length));
+    }
+
+    @Override
+    public Block getRegion(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        return createMapBlockInternal(
+                position + getOffsetBase(),
+                length,
+                Optional.ofNullable(getMapIsNull()),
+                getOffsets(),
+                getRawKeyBlock(),
+                getRawValueBlock(),
+                getHashTables());
+    }
+
+    @Override
+    public long getRegionSizeInBytes(int position, int length)
+    {
+        int positionCount = getPositionCount();
+        checkValidRegion(positionCount, position, length);
+
+        int entriesStart = getOffsets()[getOffsetBase() + position];
+        int entriesEnd = getOffsets()[getOffsetBase() + position + length];
+        int entryCount = entriesEnd - entriesStart;
+
+        return getRawKeyBlock().getRegionSizeInBytes(entriesStart, entryCount) +
+                getRawValueBlock().getRegionSizeInBytes(entriesStart, entryCount) +
+                (Integer.BYTES + Byte.BYTES) * (long) length +
+                Integer.BYTES * HASH_MULTIPLIER * (long) entryCount;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).